src/main/java/com/snowflake/kafka/connector/internal/ingestsdk/KcStreamingIngestClient.java

```diff
@@ -0,0 +1,181 @@
+   * @param clientId the client id
+   * @return the client's name as 'KC_CLIENT_kcInstanceId_clientId'
+   */
+  public static String buildStreamingIngestClientName(String kcInstanceId, int clientId) {
+    return Utils.formatString("{}_{}_{}", STREAMING_CLIENT_PREFIX_NAME, kcInstanceId, clientId);
+  }
+
+  // TESTING ONLY - inject the client
+  @VisibleForTesting
+  public KcStreamingIngestClient(SnowflakeStreamingIngestClient client) {
+    this.client = client;
+  }
+
+  /**
+   * Creates a streaming client from the given properties and requested name. Validates the
+   * requested client is not null and has the correct name
+   *
+   * <p>Any exceptions will be passed up, a sfexception will be converted to a connectexception
+   *
+   * @param streamingClientProps the properties for the client
+   * @param clientName the client name to uniquely identify the client
+   */
+  protected KcStreamingIngestClient(Properties streamingClientProps, String clientName) {
+    try {
+      LOGGER.info("Creating Streaming Client: {}", clientName);
+      this.client =
+          SnowflakeStreamingIngestClientFactory.builder(clientName)
+              .setProperties(streamingClientProps)
+              .build();
+
+      assert this.client != null; // client is final, so never need to do another null check
+      assert this.client.getName().equals(clientName);
+    } catch (SFException ex) {
+      throw new ConnectException(ex);
+    }
+  }
+
+  /**
+   * Creates an ingest sdk OpenChannelRequest and opens the client's channel
+   *
+   * <p>No exception handling done, all exceptions will be passed through
+   *
+   * @param channelName the name of the channel to open
+   * @param config config to get the database and schema names for the channel
+   * @param tableName table name of the channel
+   * @return the opened channel
+   */
+  public SnowflakeStreamingIngestChannel openChannel(
+      String channelName, Map<String, String> config, String tableName) {
+    OpenChannelRequest channelRequest =
+        OpenChannelRequest.builder(channelName)
+            .setDBName(config.get(Utils.SF_DATABASE))
+            .setSchemaName(config.get(Utils.SF_SCHEMA))
+            .setTableName(tableName)
+            .setOnErrorOption(OpenChannelRequest.OnErrorOption.CONTINUE)
+            .build();
+    LOGGER.info("Opening a channel with name:{} for table name:{}", channelName, tableName);
+
+    return this.client.openChannel(channelRequest);
+  }
+
+  /**
+   * TODO @rcheng: should add retry here and create() even if sdk retries bc network/load issues?
+   * esp with rowset api later
+   *
+   * <p>Calls the ingest sdk to close the client sdk
+   *
+   * <p>Swallows all exceptions and returns t/f if the client was closed because closing is best
+   * effort
+   *
+   * @return if the client was successfully closed
+   */
+  public boolean close() {
+    if (this.client.isClosed()) {
+      LOGGER.debug("Streaming client is already closed or null");
```

Comment:
nit: `Streaming client is already closed`
when can it be null?

Reply:
I'll remove the 'or null' log because in production it will never be null. the test constructor is the only place that it may introduce nulls